### PR TITLE
fix(gossip): validate blocks when receiving them over gossip

### PR DESCRIPTION
**Pull request by danwt**
# PR Standards

## Opening a pull request should be able to meet the following requirements

---

Close https://github.com/dymensionxyz/dymint/issues/656

<-- Briefly describe the content of this pull request -->

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: block/manager.go [hunk 1]
```diff
@@ -233,6 +233,12 @@ func (m *Manager) applyBlockCallback(event pubsub.Message) {
 	block := eventData.Block
 	commit := eventData.Commit
 
+	if err := m.validateBlock(&block, &commit); err != nil {
+		m.logger.Error("apply block callback, block not valid: dropping it", "err", err, "height", block.Header.Height)
+		/// TODO: can we take an action here such as dropping the peer / reducing their reputation?
+		return
+	}
+
```
Comment (danwt) [inline block/manager.go:241]: this is the critical change

File: block/retriever.go [hunk 2]
```diff
@@ -90,6 +90,9 @@ func (m *Manager) processNextDABatch(ctx context.Context, daMetaData *da.DASubmi
 			if block.Header.Height != m.store.NextHeight() {
 				continue
 			}
+			if err := m.validateBlock(block, batch.Commits[i]); err != nil {
+				return fmt.Errorf("validate block: height: %d: %w", block.Header.Height, err)
```
Comment (mtsitrin) [inline block/retriever.go:94]: i think its better to continue here

File: block/executor.go [hunk 3]
```diff
@@ -194,45 +181,8 @@ func (e *Executor) commit(ctx context.Context, state *types.State, block *types.
...
-	if err = commit.Validate(proposer, abciHeaderBytes); err != nil {
-		return err
-	}
-	return nil
-}
-
-// Execute executes the block and returns the ABCIResponses.
-func (e *Executor) Execute(ctx context.Context, state types.State, block *types.Block) (*tmstate.ABCIResponses, error) {
+// ExecuteValidBlock executes the block and returns the ABCIResponses.
+func (e *Executor) ExecuteValidBlock(ctx context.Context, state types.State, block *types.Block) (*tmstate.ABCIResponses, error) {
```
Comment (mtsitrin) [inline block/executor.go:185]: I think the `Valid` in the name is bit redundant.
the execution will fail and return err if execution will fail due to invalid block

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: block/retriever.go (hunk 2 above)
Comment (danwt) [reply block/retriever.go:94]: should you really get invalid blocks from the DA though?
Keeping it returning an error is essentially what we already do, making it continue is a change in functionality

File: block/executor.go (hunk 3 above)
Comment (danwt) [reply block/executor.go:185]: fixed

**Review by mtsitrin — APPROVED** [review]

**Comment by danwt** [comment]
it looks like some autorefactor used a types2 name, I'll fix it

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**